sw/device/tests/usbdev_iso_test.c

```diff
@@ -0,0 +1,281 @@
+ */
+static bool generating = true;
+
+/**
+ * Retrying is meaningless for this test since we have only Isochronous streams.
+ */
+static bool retrying = false;
+
+/**
+ * Are we expecting to receive data?
+ */
+static bool recving = true;
+
+/**
+ * Send only maximal length packets?
+ * (important for performance measurements on the USB, but obviously undesirable
+ *  for testing reliability/function)
+ */
+static bool max_packets = false;
+
+/**
+ * Number of streams to be created
+ */
+static const unsigned nstreams = NUM_STREAMS;
+
+OTTF_DEFINE_TEST_CONFIG();
+
+bool test_main(void) {
+  // Context state for streaming test
+  usb_testutils_streams_ctx_t *ctx = &stream_test;
+
+  CHECK(kDeviceType == kDeviceSimDV || kDeviceType == kDeviceSimVerilator ||
+            kDeviceType == kDeviceFpgaCw310,
+        "This test is not expected to run on platforms other than the "
+        "Verilator simulation or CW310 FPGA. It needs logic on the host side "
+        "to retrieve, scramble and return the generated byte stream");
+
+  LOG_INFO("Running USBDEV ISO Test");
+
+  // Check we can support the requested number of streams
+  CHECK(nstreams && nstreams < USBDEV_NUM_ENDPOINTS);
+
+  // Decide upon the number of bytes to be transferred for the entire test
+  uint32_t transfer_bytes = TRANSFER_BYTES_FPGA;
+  if (kDeviceType == kDeviceSimVerilator) {
+    transfer_bytes = TRANSFER_BYTES_VERILATOR;
+  }
+  transfer_bytes = (transfer_bytes + nstreams - 1) / nstreams;
+  LOG_INFO(" - %u stream(s), 0x%x bytes each", nstreams, transfer_bytes);
+
+  CHECK_DIF_OK(dif_pinmux_init(
+      mmio_region_from_addr(TOP_EARLGREY_PINMUX_AON_BASE_ADDR), &pinmux));
+  pinmux_testutils_init(&pinmux);
+  CHECK_DIF_OK(dif_pinmux_input_select(
+      &pinmux, kTopEarlgreyPinmuxPeripheralInUsbdevSense,
+      kTopEarlgreyPinmuxInselIoc7));
+
+  // Construct the test/stream flags to be used
+  test_flags = (sending ? kUsbdevStreamFlagRetrieve : 0U) |
+               (generating ? kUsbdevStreamFlagCheck : 0U) |
+               (retrying ? kUsbdevStreamFlagRetry : 0U) |
+               (recving ? kUsbdevStreamFlagSend : 0U) |
+               // Note: the 'max_packets' test flag is not required by the DPI
+               (max_packets ? kUsbdevStreamFlagMaxPackets : 0U);
+
+  // Initialize the test descriptor
+  // Note: the 'max packets' test flag is not required by the DPI model
+  const uint8_t desc[] = {USB_TESTUTILS_TEST_DSCR(
+      kUsbTestNumberIso, NUM_STREAMS | (uint8_t)test_flags, 0, 0, 0)};
+  memcpy(test_descriptor, desc, sizeof(test_descriptor));
+
+  // Remember context state for usb_testutils context
+  ctx->usbdev = &usbdev;
+
+  // Call `usbdev_init` here so that DPI will not start until the
+  // simulation has finished all of the printing, which takes a while
+  // if `--trace` was passed in.
+  CHECK_STATUS_OK(usb_testutils_init(ctx->usbdev, /*pinflip=*/false,
+                                     /*en_diff_rcvr=*/false,
```

Comment:
The CW310 board does have a transceiver with a differential receiver, and if it's disabled, the transceiver is suspended (the SUSPEND pin on the TUSB1106 would be set high). Would we not want to enable the diff receiver?

Reply:
All valid points, thank you; I'll answer them all together.
bInterval was indeed invalid for Iso streams, and it hasn't caused issues because the DPI model ignores it; in fact it presently prioritizes increased traffic/soaking over actually being Isochronous.

The build switches and confusion over which target(s) are supported are because they've been derived from usbdev_test. Until very recently usbdev_test was bypassing the differential receiver (the history of that predates my involvement), unlike hello_usbdev which did use it. All settings now revised.

I've changed usbdev_iso_test to use the receiver, but left usbdev_mixed_test bypassing it, so we're exercising both options. There is a dedicated usbdev_pincfg_test which runs through all configurations but it has not yet been merged. Both of these tests are simulation-only at present. I'll keep local mods to allow me to test FPGA builds and work on the still-incomplete host-side code for those tests.

None of the tests can yet be run in CI because we haven't merged and set up appropriate host support, and now that we have tests in the nightly regressions, doing so is less pressing. I think we should strive to sort that before the next big milestone, however.